weed/server/raft_hashicorp.go

```diff
@@ -56,46 +57,67 @@ func (s *RaftServer) AddPeersConfiguration() (cfg raft.Configuration) {
 	return cfg
 }
 
-func (s *RaftServer) UpdatePeers() {
+func (s *RaftServer) monitorLeaderLoop(updatePeers bool) {
 	for {
+		prevLeader, _ := s.RaftHashicorp.LeaderWithID()
 		select {
 		case isLeader := <-s.RaftHashicorp.LeaderCh():
+			leader, _ := s.RaftHashicorp.LeaderWithID()
 			if isLeader {
-				peerLeader := string(s.serverAddr)
-				existsPeerName := make(map[string]bool)
-				for _, server := range s.RaftHashicorp.GetConfiguration().Configuration().Servers {
-					if string(server.ID) == peerLeader {
-						continue
-					}
-					existsPeerName[string(server.ID)] = true
-				}
-				for _, peer := range s.peers {
-					peerName := string(peer)
-					if peerName == peerLeader || existsPeerName[peerName] {
-						continue
-					}
-					glog.V(0).Infof("adding new peer: %s", peerName)
-					s.RaftHashicorp.AddVoter(
-						raft.ServerID(peerName), raft.ServerAddress(peer.ToGrpcAddress()), 0, 0)
-				}
-				for peer := range existsPeerName {
-					if _, found := s.peers[peer]; !found {
-						glog.V(0).Infof("removing old peer: %s", peer)
-						s.RaftHashicorp.RemoveServer(raft.ServerID(peer), 0, 0)
-					}
+
+				if updatePeers {
+					s.updatePeers()
+					updatePeers = false
 				}
-				if _, found := s.peers[peerLeader]; !found {
-					glog.V(0).Infof("removing old leader peer: %s", peerLeader)
-					s.RaftHashicorp.RemoveServer(raft.ServerID(peerLeader), 0, 0)
+
+				glog.V(0).Infof("raft do barrier")
+				barrier := s.RaftHashicorp.Barrier(2 * time.Minute)
+				if err := barrier.Error(); err != nil {
+					glog.Fatalf("failed to wait for barrier, error %s", err)
```

Comment:
This is really not good at all

Reply:
this should not err if raft service work fine

Reply:
> this should not err if raft service work fine

I looked at the code for this function and the error can occur for several reasons. And you need to be able to correctly handle this error, since it affects only one number, maxVolumeID, and it can be double-checked. This is better than sending a message and calling to os.Exit(1)

Reply:
yes, you are right. i modify it